File: python_base_04/core/modules/recall_game/utils/yaml_rules_engine.py
"""
YAML Rules Engine - Generic interpreter for YAML-defined decision rules

Provides declarative AI decision making for computer players.
"""

import random
from typing import List, Dict, Any

class YamlRulesEngine:
    """YAML Rules Engine - Generic interpreter for YAML-defined decision rules"""
    
    def __init__(self):
        self._random = random.Random()
    
    def execute_rules(self, rules: List[Dict[str, Any]], game_data: Dict[str, Any], 
                     should_play_optimal: bool) -> str:
        """Execute YAML rules and return selected card ID"""
        # Sort rules by priority
        sorted_rules = sorted(rules, key=lambda r: r.get('priority', 999))
        
        # If not playing optimally, skip to last rule (random fallback)
        if not should_play_optimal and sorted_rules:
            last_rule = sorted_rules[-1]
            return self._execute_action(last_rule.get('action', {}), game_data)
        
        # Evaluate rules in priority order
        for rule in sorted_rules:
            condition = rule.get('condition')
            if condition and self._evaluate_condition(condition, game_data):
                action = rule.get('action')
                if action:
                    return self._execute_action(action, game_data)
        
        # Ultimate fallback: random from playable cards
        playable_cards = game_data.get('playable_cards', [])
        if playable_cards:
            return self._random.choice(playable_cards)
        
        # Last resort: random from available cards
        available_cards = game_data.get('available_cards', [])
        return self._random.choice(available_cards)
    
    def _evaluate_condition(self, condition: Dict[str, Any], game_data: Dict[str, Any]) -> bool:
        """Evaluate a condition from YAML"""
        cond_type = condition.get('type', 'always')
        
        if cond_type == 'always':
            return True
        elif cond_type == 'and':
            conditions = condition.get('conditions', [])
            return all(self._evaluate_condition(c, game_data) for c in conditions)
        elif cond_type == 'or':
            conditions = condition.get('conditions', [])
            return any(self._evaluate_condition(c, game_data) for c in conditions)
        elif cond_type == 'not':
            sub_condition = condition.get('condition')
            return not self._evaluate_condition(sub_condition, game_data) if sub_condition else False
        else:
            return self._evaluate_field_condition(condition, game_data)
    
    def _evaluate_field_condition(self, condition: Dict[str, Any], game_data: Dict[str, Any]) -> bool:
        """Evaluate a field-based condition"""
        field = condition.get('field')
        operator = condition.get('operator', 'equals')
        value = condition.get('value')
        
        if not field:
            return False
        
        field_value = game_data.get(field)
        
        if operator == 'not_empty':
            if isinstance(field_value, (list, dict)):
                return len(field_value) > 0
            return field_value is not None
        elif operator == 'empty':
            if isinstance(field_value, (list, dict)):
                return len(field_value) == 0
            return field_value is None
        elif operator == 'equals':
            return field_value == value
        elif operator == 'not_equals':
            return field_value != value
        elif operator == 'greater_than':
            return isinstance(field_value, (int, float)) and isinstance(value, (int, float)) and field_value > value
        elif operator == 'less_than':
            return isinstance(field_value, (int, float)) and isinstance(value, (int, float)) and field_value < value
        elif operator == 'contains':
            if isinstance(field_value, list):
                return value in field_value
            if isinstance(field_value, str) and isinstance(value, str):
                return value in field_value
            return False
        
        return False
    
    def _execute_action(self, action: Dict[str, Any], game_data: Dict[str, Any]) -> str:
        """Execute an action from YAML"""
        action_type = action.get('type', 'select_random')
        source = action.get('source', 'playable_cards')
        filters = action.get('filters', [])
        
        # Get source data
        source_data = list(game_data.get(source, []))
        
        # Filter out null cards from source data
        source_data = [card for card in source_data if card and str(card) != 'null']
        
        # Apply filters
        for filter_def in filters:
            if isinstance(filter_def, dict):
                source_data = self._apply_filter(source_data, filter_def, game_data)
        
        if not source_data:
            # Fallback to playable_cards
            source_data = list(game_data.get('playable_cards', []))
            source_data = [card for card in source_data if card and str(card) != 'null']
        
        if not source_data:
            # Fallback to available_cards
            source_data = list(game_data.get('available_cards', []))
            source_data = [card for card in source_data if card and str(card) != 'null']
        
        # Execute action type
        if action_type == 'select_random':
            return self._random.choice(source_data)
        elif action_type == 'select_highest_points':
            return self._select_highest_points(source_data, game_data)
        elif action_type == 'select_lowest_points':
            return self._select_lowest_points(source_data, game_data)
        elif action_type == 'select_first':
            return source_data[0]
        elif action_type == 'select_last':
            return source_data[-1]
        else:
            return self._random.choice(source_data)
    
    def _apply_filter(self, data: List[str], filter_def: Dict[str, Any], 
                     game_data: Dict[str, Any]) -> List[str]:
        """Apply a filter to source data"""
        filter_type = filter_def.get('type')
        value = filter_def.get('value')
        
        all_cards = game_data.get('all_cards_data', [])
        card_ids = set(data)
        
        if filter_type == 'exclude_rank':
            filtered = [card.get('id') for card in all_cards 
                       if card.get('id') in card_ids and card.get('rank') != value]
            return filtered
        elif filter_type == 'exclude_suit':
            filtered = [card.get('id') for card in all_cards 
                       if card.get('id') in card_ids and card.get('suit') != value]
            return filtered
        elif filter_type == 'only_rank':
            filtered = [card.get('id') for card in all_cards 
                       if card.get('id') in card_ids and card.get('rank') == value]
            return filtered
        
        return data
    
    def _select_highest_points(self, card_ids: List[str], game_data: Dict[str, Any]) -> str:
        """Select card with highest points"""
        all_cards = game_data.get('all_cards_data', [])
        card_id_set = set(card_ids)
        
        candidate_cards = [card for card in all_cards if card.get('id') in card_id_set]
        
        if not candidate_cards:
            return self._random.choice(card_ids)
        
        highest_card = max(candidate_cards, key=lambda c: c.get('points', 0))
        return highest_card.get('id', self._random.choice(card_ids))
    
    def _select_lowest_points(self, card_ids: List[str], game_data: Dict[str, Any]) -> str:
        """Select card with lowest points"""
        all_cards = game_data.get('all_cards_data', [])
        card_id_set = set(card_ids)
        
        candidate_cards = [card for card in all_cards if card.get('id') in card_id_set]
        
        if not candidate_cards:
            return self._random.choice(card_ids)
        
        lowest_card = min(candidate_cards, key=lambda c: c.get('points', 999))
        return lowest_card.get('id', self._random.choice(card_ids))

File: backups/20260119_155729_logging_optimization/dart_bkend_base_01/lib/modules/dutch_game/backend_core/shared_logic/utils/yaml_rules_engine.dart
import '../../../utils/platform/shared_imports.dart';

const bool LOGGING_SWITCH = false; // Enabled for timer-based delay system and time pressure testing

/// YAML Rules Engine - Generic interpreter for YAML-defined decision rules
class YamlRulesEngine {
  final Random _random = Random();
  final Logger _logger = Logger();
  
  /// Execute YAML rules and return selected card ID
  String executeRules(List<dynamic> rules, Map<String, dynamic> gameData, bool shouldPlayOptimal) {
    if (LOGGING_SWITCH) {
      _logger.info('Dutch: DEBUG - YamlRulesEngine.executeRules called with ${rules.length} rules, shouldPlayOptimal: $shouldPlayOptimal');
    };
    
    // Sort rules by priority
    final sortedRules = List<Map<String, dynamic>>.from(rules)
      ..sort((a, b) => (a['priority'] ?? 999).compareTo(b['priority'] ?? 999));
    
    _logger.info('Dutch: DEBUG - Sorted rules by priority: ${sortedRules.map((r) => '${r['name']} (${r['priority']})').join(', ')}', isOn: LOGGING_SWITCH);
    
    // If not playing optimally, skip to last rule (random fallback)
    if (!shouldPlayOptimal && sortedRules.isNotEmpty) {
      final lastRule = sortedRules.last;
      if (LOGGING_SWITCH) {
        _logger.info('Dutch: DEBUG - Not playing optimally, using last rule: ${lastRule['name']}');
      };
      return _executeAction(lastRule['action'], gameData);
    }
    
    // Evaluate rules in priority order
    for (final rule in sortedRules) {
      final ruleName = rule['name'] ?? 'unnamed';
      if (LOGGING_SWITCH) {
        _logger.info('Dutch: DEBUG - Evaluating rule: $ruleName');
      };
      
      final condition = rule['condition'] as Map<String, dynamic>?;
      if (condition != null) {
        final conditionResult = _evaluateCondition(condition, gameData);
        if (LOGGING_SWITCH) {
          _logger.info('Dutch: DEBUG - Rule $ruleName condition result: $conditionResult');
        };
        
        if (conditionResult) {
          final action = rule['action'] as Map<String, dynamic>?;
          if (action != null) {
            if (LOGGING_SWITCH) {
              _logger.info('Dutch: DEBUG - Rule $ruleName condition passed, executing action');
            };
            return _executeAction(action, gameData);
          }
        }
      }
    }
    
    if (LOGGING_SWITCH) {
      _logger.info('Dutch: DEBUG - No rules matched, using fallback logic');
    };
    
    // Ultimate fallback: random from playable cards
    final playableCards = gameData['playable_cards'] as List<dynamic>? ?? [];
    if (playableCards.isNotEmpty) {
      if (LOGGING_SWITCH) {
        _logger.info('Dutch: DEBUG - Using playable cards fallback: ${playableCards.length} cards');
      };
      return playableCards[_random.nextInt(playableCards.length)].toString();
    }
    
    // Last resort: random from available cards
    final availableCards = gameData['available_cards'] as List<dynamic>? ?? [];
    if (LOGGING_SWITCH) {
      _logger.info('Dutch: DEBUG - Using available cards fallback: ${availableCards.length} cards');
    };
    return availableCards[_random.nextInt(availableCards.length)].toString();
  }
  
  /// Evaluate a condition from YAML
  bool _evaluateCondition(Map<String, dynamic> condition, Map<String, dynamic> gameData) {
    final type = condition['type']?.toString() ?? 'always';
    
    switch (type) {
      case 'always':
        return true;
      
      case 'and':
        final conditions = condition['conditions'] as List<dynamic>? ?? [];
        return conditions.every((c) => _evaluateCondition(c as Map<String, dynamic>, gameData));
      
      case 'or':
        final conditions = condition['conditions'] as List<dynamic>? ?? [];
        return conditions.any((c) => _evaluateCondition(c as Map<String, dynamic>, gameData));
      
      case 'not':
        final subCondition = condition['condition'] as Map<String, dynamic>?;
        return subCondition != null ? !_evaluateCondition(subCondition, gameData) : false;
      
      default:
        // Field-based condition
        return _evaluateFieldCondition(condition, gameData);
    }
  }
  
  /// Evaluate a field-based condition
  bool _evaluateFieldCondition(Map<String, dynamic> condition, Map<String, dynamic> gameData) {
    final field = condition['field']?.toString();
    final operator = condition['operator']?.toString() ?? 'equals';
    final value = condition['value'];
    
    if (field == null) return false;
    
    final fieldValue = gameData[field];
    
    switch (operator) {
      case 'not_empty':
        if (fieldValue is List) return fieldValue.isNotEmpty;
        if (fieldValue is Map) return fieldValue.isNotEmpty;
        return fieldValue != null;
      
      case 'empty':
        if (fieldValue is List) return fieldValue.isEmpty;
        if (fieldValue is Map) return fieldValue.isEmpty;
        return fieldValue == null;
      
      case 'equals':
        return fieldValue == value;
      
      case 'not_equals':
        return fieldValue != value;
      
      case 'greater_than':
        if (fieldValue is num && value is num) return fieldValue > value;
        return false;
      
      case 'less_than':
        if (fieldValue is num && value is num) return fieldValue < value;
        return false;
      
      case 'contains':
        if (fieldValue is List) return fieldValue.contains(value);
        if (fieldValue is String && value is String) return fieldValue.contains(value);
        return false;
      
      default:
        return false;
    }
  }
  
  /// Execute an action from YAML
  String _executeAction(Map<String, dynamic> action, Map<String, dynamic> gameData) {
    final type = action['type']?.toString() ?? 'select_random';
    final source = action['source']?.toString() ?? 'playable_cards';
    final filters = action['filters'] as List<dynamic>? ?? [];
    
    if (LOGGING_SWITCH) {
      _logger.info('Dutch: DEBUG - _executeAction called with type: $type, source: $source, filters: ${filters.length}');
    };
    
    // Get source data
    List<dynamic> sourceData = gameData[source] as List<dynamic>? ?? [];
    if (LOGGING_SWITCH) {
      _logger.info('Dutch: DEBUG - Initial source data from $source: ${sourceData.length} items');
    };
    
    // CRITICAL: Always filter out null cards and collection cards from any source
    // Get collection card IDs from gameData
    final collectionCardIds = (gameData['collection_cards'] as List<dynamic>? ?? [])
        .map((id) => id.toString())
        .where((id) => id.isNotEmpty)
        .toSet();
    
    // Filter out nulls and collection cards
    sourceData = sourceData
        .where((cardId) {
          final cardIdStr = cardId.toString();
          // Filter out null cards (null, 'null', or empty strings)
          if (cardIdStr.isEmpty || cardIdStr == 'null') {
            return false;
          }
          // Filter out collection rank cards
          if (collectionCardIds.contains(cardIdStr)) {
            return false;
          }
          return true;
        })
        .toList();
    
    if (LOGGING_SWITCH) {
      _logger.info('Dutch: DEBUG - Source data after null/collection filtering: ${sourceData.length} items');
    };
    
    // Apply filters
    for (final filter in filters) {
      if (filter is Map<String, dynamic>) {
        if (LOGGING_SWITCH) {
          _logger.info('Dutch: DEBUG - Applying filter: ${filter['type']}');
        };
        sourceData = _applyFilter(sourceData, filter, gameData);
        if (LOGGING_SWITCH) {
          _logger.info('Dutch: DEBUG - After filter, source data: ${sourceData.length} items');
        };
      }
    }
    
    if (sourceData.isEmpty) {
      // Fallback to playable cards
      sourceData = gameData['playable_cards'] as List<dynamic>? ?? [];
      if (LOGGING_SWITCH) {
        _logger.info('Dutch: DEBUG - Source data empty, using playable cards fallback: ${sourceData.length} items');
      };
      
      // Filter fallback data too
      sourceData = sourceData
          .where((cardId) {
            final cardIdStr = cardId.toString();
            if (cardIdStr.isEmpty || cardIdStr == 'null') return false;
            if (collectionCardIds.contains(cardIdStr)) return false;
            return true;
          })
          .toList();
    }
    
    if (sourceData.isEmpty) {
      // Last resort: available cards
      sourceData = gameData['available_cards'] as List<dynamic>? ?? [];
      if (LOGGING_SWITCH) {
        _logger.info('Dutch: DEBUG - Still empty, using available cards fallback: ${sourceData.length} items');
      };
      
      // Filter fallback data too (but this shouldn't be used if playable_cards exists)
      sourceData = sourceData
          .where((cardId) {
            final cardIdStr = cardId.toString();
            if (cardIdStr.isEmpty || cardIdStr == 'null') return false;
            if (collectionCardIds.contains(cardIdStr)) return false;
            return true;
          })
          .toList();
    }
    
    if (sourceData.isEmpty) {
      if (LOGGING_SWITCH) {
        _logger.warning('Dutch: DEBUG - All fallbacks failed, no cards available!');
      };
      return ''; // Return empty string to indicate no card available
    }
    
    // Execute action type
    String result;
    switch (type) {
      case 'select_random':
        result = sourceData[_random.nextInt(sourceData.length)].toString();
        if (LOGGING_SWITCH) {
          _logger.info('Dutch: DEBUG - select_random result: $result');
        };
        break;
      
      case 'select_highest_points':
        result = _selectHighestPoints(sourceData, gameData);
        if (LOGGING_SWITCH) {
          _logger.info('Dutch: DEBUG - select_highest_points result: $result');
        };
        break;
      
      case 'select_lowest_points':
        result = _selectLowestPoints(sourceData, gameData);
        if (LOGGING_SWITCH) {
          _logger.info('Dutch: DEBUG - select_lowest_points result: $result');
        };
        break;
      
      case 'select_first':
        result = sourceData.first.toString();
        if (LOGGING_SWITCH) {
          _logger.info('Dutch: DEBUG - select_first result: $result');
        };
        break;
      
      case 'select_last':
        result = sourceData.last.toString();
        if (LOGGING_SWITCH) {
          _logger.info('Dutch: DEBUG - select_last result: $result');
        };
        break;
      
      default:
        result = sourceData[_random.nextInt(sourceData.length)].toString();
        if (LOGGING_SWITCH) {
          _logger.info('Dutch: DEBUG - default action result: $result');
        };
        break;
    }
    
    return result;
  }
  
  /// Apply a filter to source data
  List<dynamic> _applyFilter(List<dynamic> data, Map<String, dynamic> filter, Map<String, dynamic> gameData) {
    final type = filter['type']?.toString();
    final value = filter['value'];
    
    switch (type) {
      case 'exclude_rank':
        // Filter out cards with specific rank
        final allCards = gameData['all_cards_data'] as List<dynamic>? ?? [];
        final cardIds = data.map((d) => d.toString()).toSet();
        final filteredCards = allCards.where((card) {
          if (card is Map<String, dynamic>) {
            final cardId = card['id']?.toString();
            final rank = card['rank']?.toString();
            return cardIds.contains(cardId) && rank != value;
          }
          return false;
        }).map((card) => (card as Map<String, dynamic>)['id']).toList();
        return filteredCards;
      
      case 'exclude_suit':
        final allCards = gameData['all_cards_data'] as List<dynamic>? ?? [];
        final cardIds = data.map((d) => d.toString()).toSet();
        final filteredCards = allCards.where((card) {
          if (card is Map<String, dynamic>) {
            final cardId = card['id']?.toString();
            final suit = card['suit']?.toString();
            return cardIds.contains(cardId) && suit != value;
          }
          return false;
        }).map((card) => (card as Map<String, dynamic>)['id']).toList();
        return filteredCards;
      
      case 'only_rank':
        final allCards = gameData['all_cards_data'] as List<dynamic>? ?? [];
        final cardIds = data.map((d) => d.toString()).toSet();
        final filteredCards = allCards.where((card) {
          if (card is Map<String, dynamic>) {
            final cardId = card['id']?.toString();
            final rank = card['rank']?.toString();
            return cardIds.contains(cardId) && rank == value;
          }
          return false;
        }).map((card) => (card as Map<String, dynamic>)['id']).toList();
        return filteredCards;
      
      default:
        return data;
    }
  }
  
  /// Select card with highest points
  String _selectHighestPoints(List<dynamic> cardIds, Map<String, dynamic> gameData) {
    final allCards = gameData['all_cards_data'] as List<dynamic>? ?? [];
    final cardIdSet = cardIds.map((d) => d.toString()).toSet();
    
    final candidateCards = allCards.where((card) {
      if (card is Map<String, dynamic>) {
        final cardId = card['id']?.toString();
        return cardIdSet.contains(cardId);
      }
      return false;
    }).cast<Map<String, dynamic>>().toList();
    
    if (candidateCards.isEmpty) {
      return cardIds[_random.nextInt(cardIds.length)].toString();
    }
    
    Map<String, dynamic>? highestCard;
    int highestPoints = -1;
    
    for (final card in candidateCards) {
      final points = card['points'] as int? ?? 0;
      if (points > highestPoints) {
        highestPoints = points;
        highestCard = card;
      }
    }
    
    return highestCard?['id']?.toString() ?? cardIds[_random.nextInt(cardIds.length)].toString();
  }
  
  /// Select card with lowest points
  String _selectLowestPoints(List<dynamic> cardIds, Map<String, dynamic> gameData) {
    final allCards = gameData['all_cards_data'] as List<dynamic>? ?? [];
    final cardIdSet = cardIds.map((d) => d.toString()).toSet();
    
    final candidateCards = allCards.where((card) {
      if (card is Map<String, dynamic>) {
        final cardId = card['id']?.toString();
        return cardIdSet.contains(cardId);
      }
      return false;
    }).cast<Map<String, dynamic>>().toList();
    
    if (candidateCards.isEmpty) {
      return cardIds[_random.nextInt(cardIds.length)].toString();
    }
    
    Map<String, dynamic>? lowestCard;
    int lowestPoints = 999;
    
    for (final card in candidateCards) {
      final points = card['points'] as int? ?? 0;
      if (points < lowestPoints) {
        lowestPoints = points;
        lowestCard = card;
      }
    }
    
    return lowestCard?['id']?.toString() ?? cardIds[_random.nextInt(cardIds.length)].toString();
  }
}

